File: python/packages/var-core/src/var_core/conformance.py
"""conformance.py — var-doc, registry, plan, and trace artifact projections.

Port of toVarDocArtifact, toRegistryArtifact, toPlanArtifact, toFailureArtifact,
and runConformance from typescript/packages/var-core/src/conformance.ts.
Serializes a VarDoc AST / Registry / ExecutionPlan / trace to the camelCase wire
dicts expected by the conformance golden files.
"""

from __future__ import annotations

import os
from dataclasses import dataclass
from typing import Any, Callable

from cucumber_expressions.expression import CucumberExpression

from var_core.ast import (
    Blockquote,
    Example,
    Fence,
    Heading,
    InlineOffset,
    ListItem,
    Paragraph,
    Row,
    Table,
    ThematicBreak,
    VarDoc,
)
from var_core.cell_diff import ReturnShapeError, is_cell_mismatch_error
from var_core.doc_string_diff import is_doc_string_mismatch_error
from var_core.execute import CollectPorts, StepObservation, collect_examples, is_unexpected_pass_error
from var_core.plan import ExecutionPlan
from var_core.plan import plan as build_plan
from var_core.registry import Registry
from var_core.span import Span, utf16_slice


@dataclass(frozen=True, slots=True)
class BundleArtifacts:
    """Typed container for the four-artifact conformance bundle."""

    var_doc: dict[str, Any]
    registry: dict[str, Any]
    plan: dict[str, Any]
    trace: dict[str, Any]


# ---------------------------------------------------------------------------
# Registry artifact projection
# ---------------------------------------------------------------------------


def parameter_type_names(compiled: CucumberExpression) -> list[str]:
    """Return parameter-type names in source order from a compiled expression.

    Port of ``parameterTypeNames`` in conformance.ts.  The TS implementation
    walks ``compiled.ast`` collecting ``NodeType.parameter`` nodes; in Python
    ``CucumberExpression`` does not expose ``.ast``, but ``compiled.parameter_types``
    is populated in source order by ``rewrite_parameter`` during ``__init__``,
    giving identical results.
    """
    return [pt.name for pt in compiled.parameter_types]


def to_registry_artifact(
    registry: Registry,
    parameter_types: list[dict[str, str]] | None = None,
) -> dict[str, Any]:
    """Project a Registry to the camelCase wire dict for the registry artifact.

    Port of ``toRegistryArtifact`` from conformance.ts.
    """
    if parameter_types is None:
        parameter_types = []
    return {
        "steps": [
            {
                "expression": s.expression,
                "parameterTypeNames": parameter_type_names(s.compiled),
            }
            for s in registry.steps
        ],
        "parameterTypes": [
            {"name": p["name"], "regexp": p["regexp"]} for p in parameter_types
        ],
    }


def _span(s: Span) -> dict[str, Any]:
    return {
        "startOffset": s.start_offset,
        "endOffset": s.end_offset,
        "startLine": s.start_line,
        "startCol": s.start_col,
        "endLine": s.end_line,
        "endCol": s.end_col,
    }


def _inline(io: InlineOffset) -> dict[str, Any]:
    return {
        "textOffset": io.text_offset,
        "sourceOffset": io.source_offset,
    }


def _row(r: Row) -> dict[str, Any]:
    return {
        "cells": list(r.cells),
        "cellSpans": [_span(cs) for cs in r.cell_spans],
        "span": _span(r.span),  # type: ignore[arg-type]
    }


def _block(b: Heading | Paragraph | ListItem | Blockquote | Table | Fence | ThematicBreak) -> dict[str, Any]:
    if isinstance(b, Paragraph):
        return {
            "kind": b.kind,
            "text": b.text,
            "span": _span(b.span),  # type: ignore[arg-type]
            "inlineMap": [_inline(io) for io in b.inline_map],
        }
    if isinstance(b, Heading):
        return {
            "kind": b.kind,
            "level": b.level,
            "text": b.text,
            "span": _span(b.span),  # type: ignore[arg-type]
        }
    if isinstance(b, ListItem):
        return {
            "kind": b.kind,
            "text": b.text,
            "span": _span(b.span),  # type: ignore[arg-type]
            "inlineMap": [_inline(io) for io in b.inline_map],
            "ordered": b.ordered,
            "markerSpan": _span(b.marker_span),  # type: ignore[arg-type]
        }
    if isinstance(b, Blockquote):
        return {
            "kind": b.kind,
            "text": b.text,
            "span": _span(b.span),  # type: ignore[arg-type]
            "inlineMap": [_inline(io) for io in b.inline_map],
        }
    if isinstance(b, Table):
        return {
            "kind": b.kind,
            "span": _span(b.span),  # type: ignore[arg-type]
            "header": _row(b.header),  # type: ignore[arg-type]
            "rows": [_row(r) for r in b.rows],
        }
    if isinstance(b, Fence):
        return {
            "kind": b.kind,
            "span": _span(b.span),  # type: ignore[arg-type]
            "info": b.info,
            "body": b.body,
            "bodySpan": _span(b.body_span),  # type: ignore[arg-type]
        }
    if isinstance(b, ThematicBreak):
        return {
            "kind": b.kind,
            "span": _span(b.span),  # type: ignore[arg-type]
        }
    raise TypeError(f"Unknown block type: {type(b)}")  # pragma: no cover


def _example(ex: Example) -> dict[str, Any]:
    return {
        "scopeStack": list(ex.scope_stack),
        "span": _span(ex.span),  # type: ignore[arg-type]
        "body": [_block(b) for b in ex.body],
    }


def to_var_doc_artifact(doc: VarDoc) -> dict[str, Any]:
    """Project a VarDoc to the camelCase wire dict for the var-doc artifact."""
    return {
        "path": doc.path,
        "examples": [_example(ex) for ex in doc.examples],
        "orphanAttachments": [_block(b) for b in doc.orphan_attachments],
    }


# ---------------------------------------------------------------------------
# Plan artifact projection
# ---------------------------------------------------------------------------


def _doc_string(ds: Any) -> dict[str, Any]:
    return {
        "content": ds.content,
        "contentType": ds.content_type,
        "span": _span(ds.span),
    }


def to_plan_artifact(plan: ExecutionPlan) -> dict[str, Any]:
    """Project an ExecutionPlan to the camelCase wire dict for the plan artifact.

    Port of ``toPlanArtifact`` from conformance.ts.
    """
    source = plan.var_doc.source

    def _step(step: Any) -> dict[str, Any]:
        step_names = parameter_type_names(step.step_def.compiled)
        result: dict[str, Any] = {
            "text": step.text,
            "matchSpan": _span(step.match_span),
            "paramSpans": [_span(s) for s in step.param_spans],
            "matchedExpression": step.step_def.expression,
            "args": [
                {
                    "value": utf16_slice(source, s.start_offset, s.end_offset),
                    "parameterType": step_names[i] if i < len(step_names) else None,
                }
                for i, s in enumerate(step.param_spans)
            ],
        }
        if step.data_table is not None:
            result["dataTable"] = _block(step.data_table)
        if step.doc_string is not None:
            result["docString"] = _doc_string(step.doc_string)
        return result

    def _planned_example(ex: Any) -> dict[str, Any]:
        result: dict[str, Any] = {
            "name": ex.name,
            "scopeStack": list(ex.scope_stack),
            "span": _span(ex.span),
            "expectedOutcome": ex.expected_outcome if ex.expected_outcome is not None else "pass",
        }
        if ex.expected_error_message is not None:
            result["expectedErrorMessage"] = ex.expected_error_message
        result["steps"] = [_step(s) for s in ex.steps]
        return result

    return {
        "examples": [_planned_example(ex) for ex in plan.examples],
        "diagnostics": [
            {
                "code": d.code,
                "severity": d.severity,
                "span": _span(d.span),
            }
            for d in plan.diagnostics
        ],
    }


# ---------------------------------------------------------------------------
# Trace artifact projection
# ---------------------------------------------------------------------------


def _file_stem(path: str) -> str:
    """Return the file stem: ``path/to/foo.steps.py`` -> ``foo.steps``.

    Port of ``fileStem`` from conformance.ts.  Strips the final extension so
    that ``foo.steps.py`` becomes ``foo.steps`` and ``foo.py`` becomes ``foo``.
    """
    base = os.path.basename(path)
    # Strip the last extension only (same as TS: base.replace(/\\.[^.]+$/, ''))
    stem, _ext = os.path.splitext(base)
    return stem


def to_failure_artifact(error: object, line: int) -> dict[str, Any]:
    """Project an execution error to a FailureArtifact dict.

    Port of ``toFailureArtifact`` from conformance.ts.
    ``line`` is the failing step's ``match_span.start_line`` (1-based).
    """
    if is_cell_mismatch_error(error):
        return {
            "kind": "cell-mismatch",
            "line": line,
            "cells": [
                {
                    "column": c.column,
                    "expected": c.expected,
                    "actual": c.actual,
                    "span": _span(c.span),
                }
                for c in error.cells
                if not c.ok
            ],
        }
    if is_doc_string_mismatch_error(error):
        return {
            "kind": "doc-string-mismatch",
            "line": line,
            "diff": {
                "expected": error.diff.expected,
                "actual": error.diff.actual,
                "span": _span(error.diff.span),
            },
        }
    if isinstance(error, ReturnShapeError):
        return {"kind": "return-shape", "line": line}
    if is_unexpected_pass_error(error):
        return {"kind": "unexpected-pass", "line": line}
    return {"kind": "thrown", "line": line}


def run_conformance(
    var_doc: VarDoc,
    registry: Registry,
    create_context: Callable[[str], Any],
    parameter_types: tuple[dict[str, str], ...] = (),
) -> BundleArtifacts:
    """Run all examples and return the four-artifact bundle as a typed BundleArtifacts.

    Port of ``runConformance`` from conformance.ts.
    """
    execution = build_plan(var_doc, registry)

    # Accumulate step observations keyed by example index.
    observed: dict[int, list[StepObservation]] = {}

    class _RecordingObserver:
        def step(self, o: StepObservation) -> None:
            lst = observed.setdefault(o.example_index, [])
            lst.append(o)

    class _NullReporter:
        def diagnostic(self, _d: Any) -> None:
            pass

    queue = collect_examples(
        execution,
        CollectPorts(
            reporter=_NullReporter(),
            create_context=create_context,
            observer=_RecordingObserver(),
        ),
    )

    trace_examples = []
    for k, queued in enumerate(queue):
        outcome: str = "pass"
        try:
            queued.run()
        except Exception:
            outcome = "fail"

        planned = execution.examples[k]
        obs_list = observed.get(k, [])

        steps = []
        for i, step in enumerate(planned.steps):
            # Select the fail observation if any, else the last; skipped if none.
            ordinal = i + 1
            matches = [x for x in obs_list if x.ordinal == ordinal]
            o = next((m for m in matches if m.outcome == "fail"), None)
            if o is None and matches:
                o = matches[-1]

            step_outcome: str = o.outcome if o is not None else "skipped"
            step_dict: dict[str, Any] = {
                "exampleName": queued.name,
                "ordinal": ordinal,
                "stepText": step.text,
                "matchedExpression": step.step_def.expression,
                "contextKey": {
                    "exampleName": queued.name,
                    "stepFile": _file_stem(step.step_def.expression_source_file),
                },
                "outcome": step_outcome,
            }
            if step_outcome == "fail":
                step_dict["failure"] = to_failure_artifact(
                    o.error if o is not None else None,
                    step.match_span.start_line,
                )
            steps.append(step_dict)

        trace_examples.append({"name": queued.name, "outcome": outcome, "steps": steps})

    return BundleArtifacts(
        var_doc=to_var_doc_artifact(var_doc),
        registry=to_registry_artifact(registry, list(parameter_types)),
        plan=to_plan_artifact(execution),
        trace={"examples": trace_examples},
    )

File: dotnet/Varar.Core/Conformance.cs
using System.Collections.Immutable;

namespace Varar.Core;

/// <summary>
/// Projects pipeline output to the shared conformance wire shapes (serialized by
/// <see cref="CanonicalJson"/>). Port of the projections in <c>conformance.ts</c>. This file grows
/// one projection per artifact as the pipeline lands; T3 covers <c>var-doc.json</c>.
/// </summary>
public static class Conformance
{
    /// <summary>
    /// Projects a <see cref="Registry"/> to the <c>registry.json</c> shape: each step's expression
    /// plus its parameter-type names (in source order, from the compiled expression), and the
    /// custom parameter types as <c>{name, regexp}</c>. Port of <c>toRegistryArtifact</c>.
    /// </summary>
    public static Value ToRegistryArtifact(Registry registry) => Map(
        ("steps", Value.List(registry.Steps.Select(step => Map(
            ("expression", Value.Of(step.Expression)),
            ("parameterTypeNames", Value.List(step.Compiled.ParameterTypes.Select(p => Value.Of(p.Name)))))))),
        ("parameterTypes", Value.List(registry.CustomParameterTypes.Select(custom => Map(
            ("name", Value.Of(custom.Name)),
            ("regexp", Value.Of(custom.Regexp)))))));

    /// <summary>
    /// Projects an <see cref="ExecutionPlan"/> to the <c>plan.json</c> shape: per example the name,
    /// scope, span, expected outcome, and steps; each step's text/spans/expression, its args as the
    /// raw source slice of each param span plus its type name, and any attached table/doc string.
    /// Port of <c>toPlanArtifact</c>.
    /// </summary>
    public static Value ToPlanArtifact(ExecutionPlan plan) => Map(
        ("examples", Value.List(plan.Examples.Select(ex => PlannedExampleValue(ex, plan.VarDoc.Source)))),
        ("diagnostics", Value.List(plan.Diagnostics.Select(d => Map(
            ("code", Value.Of(d.Code.ToWire())),
            ("severity", Value.Of(d.Severity.ToWire())),
            ("span", SpanValue(d.Span)))))));

    /// <summary>
    /// Runs the plan and projects the recorded events to the <c>trace.json</c> shape: per example the
    /// name/outcome and per-step ordinal/text/expression/context-key/outcome, with a structured
    /// failure artifact on failure. Built inline (as TS's <c>runConformance</c> does).
    /// </summary>
    public static Value ToTraceArtifact(ExecutionPlan plan, Func<string, Value> createContext)
    {
        var traceExamples = plan.Examples.Select(ex =>
        {
            var observations = new List<StepObservation>();
            var failure = Execute.RunExample(plan, ex, createContext, observations);
            var outcome = failure is null ? "pass" : "fail";

            var steps = ex.Steps.Select((step, i) =>
            {
                var matches = observations.Where(o => o.Ordinal == i + 1).ToList();
                var observed = matches.FirstOrDefault(m => m.Outcome == "fail")
                    ?? (matches.Count > 0 ? matches[^1] : null);
                var stepOutcome = observed?.Outcome ?? "skipped";

                var entries = new List<KeyValuePair<string, Value>>
                {
                    new("exampleName", Value.Of(ex.Name)),
                    new("ordinal", Value.Of(i + 1)),
                    new("stepText", Value.Of(step.Text)),
                    new("matchedExpression", Value.Of(step.StepDef.Expression)),
                    new("contextKey", Map(
                        ("exampleName", Value.Of(ex.Name)),
                        ("stepFile", Value.Of(FileStem(step.StepDef.ExpressionSourceFile))))),
                    new("outcome", Value.Of(stepOutcome)),
                };
                if (stepOutcome == "fail")
                {
                    entries.Add(new("failure", ToFailureArtifact(observed?.Error, step.MatchSpan)));
                }

                return Value.Map(entries);
            });

            return Map(("name", Value.Of(ex.Name)), ("outcome", Value.Of(outcome)), ("steps", Value.List(steps)));
        });

        return Map(("examples", Value.List(traceExamples)));
    }

    private static Value ToFailureArtifact(Exception? error, Span matchSpan)
    {
        var entries = new List<KeyValuePair<string, Value>>
        {
            new("line", Value.Of(matchSpan.StartLine)),
            new("anchor", SpanValue(FailureAnchor.Anchor(error, matchSpan))),
        };
        switch (error)
        {
            case CellMismatchError cm:
                entries.Add(new("kind", Value.Of("cell-mismatch")));
                entries.Add(new("cells", Value.List(cm.Cells.Where(c => !c.Ok).Select(c => Map(
                    ("column", Value.Of(c.Column)),
                    ("expected", Value.Of(c.Expected)),
                    ("actual", Value.Of(c.Actual)),
                    ("span", SpanValue(c.Span)))))));
                break;
            case DocStringMismatchError dm:
                entries.Add(new("kind", Value.Of("doc-string-mismatch")));
                entries.Add(new("diff", Map(
                    ("expected", Value.Of(dm.Diff.Expected)),
                    ("actual", Value.Of(dm.Diff.Actual)),
                    ("span", SpanValue(dm.Diff.Span)))));
                break;
            case ReturnShapeError:
                entries.Add(new("kind", Value.Of("return-shape")));
                break;
            case UnexpectedPassError:
                entries.Add(new("kind", Value.Of("unexpected-pass")));
                break;
            default:
                entries.Add(new("kind", Value.Of("thrown")));
                break;
        }

        return Value.Map(entries);
    }

    // A step-def file's stem — the cross-port stepFile (numerals.steps.cs → numerals.steps).
    private static string FileStem(string path) => Path.GetFileNameWithoutExtension(path);

    private static Value PlannedExampleValue(PlannedExample ex, string source)
    {
        var entries = new List<KeyValuePair<string, Value>>
        {
            new("name", Value.Of(ex.Name)),
            new("scopeStack", Value.List(ex.ScopeStack.Select(Value.Of))),
            new("span", SpanValue(ex.Span)),
            new("expectedOutcome", Value.Of(ex.ExpectedFail ? "fail" : "pass")),
        };
        if (ex.ExpectedErrorMessage is not null)
        {
            entries.Add(new("expectedErrorMessage", Value.Of(ex.ExpectedErrorMessage)));
        }

        entries.Add(new("steps", Value.List(ex.Steps.Select(step => PlannedStepValue(step, source)))));
        return Value.Map(entries);
    }

    private static Value PlannedStepValue(PlannedStep step, string source)
    {
        var typeNames = step.StepDef.Compiled.ParameterTypes.Select(p => p.Name).ToArray();
        var entries = new List<KeyValuePair<string, Value>>
        {
            new("text", Value.Of(step.Text)),
            new("matchSpan", SpanValue(step.MatchSpan)),
            new("paramSpans", List(step.ParamSpans, SpanValue)),
            new("matchedExpression", Value.Of(step.StepDef.Expression)),
            new("args", Value.List(step.ParamSpans.Select((span, i) => Map(
                ("value", Value.Of(Scanner.Slice(source, span.StartOffset, span.EndOffset))),
                ("parameterType", i < typeNames.Length ? Value.Of(typeNames[i]) : Value.Null))))),
        };
        if (step.DataTable is not null)
        {
            entries.Add(new("dataTable", BlockValue(step.DataTable)));
        }

        if (step.DocString is not null)
        {
            entries.Add(new("docString", Map(
                ("content", Value.Of(step.DocString.Content)),
                ("contentType", Value.Of(step.DocString.ContentType)),
                ("span", SpanValue(step.DocString.Span)))));
        }

        return Value.Map(entries);
    }

    /// <summary>Projects a <see cref="VarDoc"/> to the <c>var-doc.json</c> shape (<c>source</c> is dropped).</summary>
    public static Value ToVarDocArtifact(VarDoc doc) => Map(
        ("path", Value.Of(doc.Path)),
        ("examples", List(doc.Examples, ExampleValue)),
        ("orphanAttachments", List(doc.OrphanAttachments, BlockValue)));

    private static Value ExampleValue(Example example) => Map(
        ("scopeStack", Value.List(example.ScopeStack.Select(Value.Of))),
        ("span", SpanValue(example.Span)),
        ("body", List(example.Body, BlockValue)));

    private static Value BlockValue(Block block) => block switch
    {
        Heading h => Map(
            ("kind", Value.Of(h.Kind)),
            ("level", Value.Of(h.Level)),
            ("text", Value.Of(h.Text)),
            ("span", SpanValue(h.Span))),
        Paragraph p => Map(
            ("kind", Value.Of(p.Kind)),
            ("text", Value.Of(p.Text)),
            ("span", SpanValue(p.Span)),
            ("segmentMap", List(p.SegmentMap, SegmentValue))),
        ListItem li => Map(
            ("kind", Value.Of(li.Kind)),
            ("ordered", Value.Of(li.Ordered)),
            ("text", Value.Of(li.Text)),
            ("span", SpanValue(li.Span)),
            ("segmentMap", List(li.SegmentMap, SegmentValue)),
            ("markerSpan", SpanValue(li.MarkerSpan))),
        Blockquote bq => Map(
            ("kind", Value.Of(bq.Kind)),
            ("text", Value.Of(bq.Text)),
            ("span", SpanValue(bq.Span)),
            ("segmentMap", List(bq.SegmentMap, SegmentValue))),
        Table t => Map(
            ("kind", Value.Of(t.Kind)),
            ("span", SpanValue(t.Span)),
            ("header", RowValue(t.Header)),
            ("rows", List(t.Rows, RowValue))),
        Fence f => Map(
            ("kind", Value.Of(f.Kind)),
            ("info", Value.Of(f.Info)),
            ("body", Value.Of(f.Body)),
            ("bodySpan", SpanValue(f.BodySpan)),
            ("span", SpanValue(f.Span))),
        ThematicBreak tb => Map(
            ("kind", Value.Of(tb.Kind)),
            ("span", SpanValue(tb.Span))),
        _ => throw new InvalidOperationException($"unknown block kind: {block.Kind}"),
    };

    private static Value RowValue(Row row) => Map(
        ("cells", Value.List(row.Cells.Select(Value.Of))),
        ("cellSpans", List(row.CellSpans, SpanValue)),
        ("span", SpanValue(row.Span)));

    private static Value SegmentValue(SegmentOffset segment) => Map(
        ("textOffset", Value.Of(segment.TextOffset)),
        ("sourceOffset", Value.Of(segment.SourceOffset)));

    private static Value SpanValue(Span span) => Map(
        ("startOffset", Value.Of(span.StartOffset)),
        ("endOffset", Value.Of(span.EndOffset)),
        ("startLine", Value.Of(span.StartLine)),
        ("startCol", Value.Of(span.StartCol)),
        ("endLine", Value.Of(span.EndLine)),
        ("endCol", Value.Of(span.EndCol)));

    private static Value List<T>(ImmutableArray<T> items, Func<T, Value> project) =>
        Value.List(items.Select(project));

    private static Value Map(params (string Key, Value Value)[] entries) =>
        Value.Map(entries.Select(e => new KeyValuePair<string, Value>(e.Key, e.Value)));
}
